### PR TITLE
KAFKA-18027: MINOR: Correct DelayedOperationPurgatory code around adding of an already completed operation

### PR DESCRIPTION
**Pull request by adixitconfluent**
### About
In reference to PR comment https://github.com/apache/kafka/pull/17739#discussion_r1844998612 ,the code seems inaccurate and not aligned with the below comment. I've made the code change to align it with the comment

> It is possible that an operation is completed after it has been added to the watch list for some, but not all the keys. In this case, the operation is considered completed and won't be added to the watch list of the remaining keys.

### CODE REVIEW

**Review by apoorvmittal10 — APPROVED** [review]

File: server-common/src/main/java/org/apache/kafka/server/purgatory/DelayedOperationPurgatory.java [hunk 1]
```diff
@@ -105,8 +105,8 @@ private WatcherList watcherList(DelayedOperationKey key) {
      * <br/>
      * Note that a delayed operation can be watched on multiple keys. It is possible that
      * an operation is completed after it has been added to the watch list for some, but
-     * not all the keys. In this case, the operation is considered completed and won't
-     * be added to the watch list of the remaining keys. The expiration reaper thread will
+     * not all the keys. In this case, although the operation will be added for the remaining keys, it is considered
+     * completed and won't be watched for the remaining keys. The expiration reaper thread will
```
Comment (junrao) [inline server-common/src/main/java/org/apache/kafka/server/purgatory/DelayedOperationPurgatory.java]: > won't be watched for the remaining keys

The code doesn't do that though. In the following, `watchForOperation` doesn't check if the operation is completed or not. So, we need to either change the code or the comment.

```
        if (operation.safeTryCompleteOrElse(() -> {
            watchKeys.forEach(key -> watchForOperation(key, operation));
            if (!watchKeys.isEmpty())
                estimatedTotalOperations.incrementAndGet();
        })) {
            return true;
        }

```

**Review by junrao — COMMENTED** [review]
@adixitconfluent : Thanks for the PR. Left a comment.

File: server-common/src/main/java/org/apache/kafka/server/purgatory/DelayedOperationPurgatory.java (hunk 1 above)
Comment (adixitconfluent) [reply server-common/src/main/java/org/apache/kafka/server/purgatory/DelayedOperationPurgatory.java]: Hi @junrao, making the code change made more sense to me. Hence, I've updated the code.

Comment (junrao) [reply server-common/src/main/java/org/apache/kafka/server/purgatory/DelayedOperationPurgatory.java]: > although the operation will be added for the remaining keys

The comment is no longer accurate with the new code.

**Review by junrao — COMMENTED** [review]
@adixitconfluent : Thanks for the updated PR. Left a minor comment.

**Comment by adixitconfluent** [comment]
Hi @junrao , my bad, thanks for pointing it out, I have removed the comment change. Now, it seems the code is aligned with the comment. Please review when you can. Thanks!

**Review by junrao — COMMENTED** [review]
@adixitconfluent : Thanks for the updated PR. The code LGTM. Could you change the description of the PR and jira accordingly?

**Comment by adixitconfluent** [comment]
> @adixitconfluent : Thanks for the updated PR. The code LGTM. Could you change the description of the PR and jira accordingly?

Done @junrao , thanks

**Comment by junrao** [comment]
@adixitconfluent : Are the test failures related to the PR?

**Comment by adixitconfluent** [comment]
> @adixitconfluent : Are the test failures related to the PR?

Hi @junrao , no, I don't think the test failure if related to my changes. I ran the test on my branch locally multiple times, it ran correctly all the times. This is an already known flaky test https://issues.apache.org/jira/browse/KAFKA-18038 .

**Review by junrao — APPROVED** [review]
@adixitconfluent : Thanks for triaging the test failures. LGTM

**Pull request merged**